core/src/tile/tileManager.cpp

```diff
@@ -303,204 +215,244 @@ void TileManager::updateTileSet(TileSet& tileSet) {
 }
 
-void TileManager::enqueueTask(const TileSet& tileSet, const TileID& tileID,
-                              const glm::dvec2& viewCenter) {
+void TileManager::enqueueTask(TileSet& _tileSet, const TileID& _tileID,
+                              const glm::dvec2& _viewCenter) {
 
     // Keep the items sorted by distance
-    auto tileCenter = m_view->getMapProjection().TileCenter(tileID);
-    double distance = glm::length2(tileCenter - viewCenter);
+    auto tileCenter = m_view->getMapProjection().TileCenter(_tileID);
+    double distance = glm::length2(tileCenter - _viewCenter);
 
     auto it = std::upper_bound(m_loadTasks.begin(), m_loadTasks.end(), distance,
                                [](auto& distance, auto& other){
                                    return distance < std::get<0>(other);
                                });
 
-    m_loadTasks.insert(it, std::make_tuple(distance, &tileSet, &tileID));
+    m_loadTasks.insert(it, std::make_tuple(distance, &_tileSet, &_tileID));
 }
 
 void TileManager::loadTiles() {
 
-    for (auto& item : m_loadTasks) {
+    for (auto& loadTask : m_loadTasks) {
 
-        auto id = *std::get<2>(item);
-        auto& tileSet = *std::get<1>(item);
-        auto tileIt = tileSet.tiles.find(id);
-        auto tile = tileIt->second;
-        auto task = std::make_shared<TileTask>(tile, tileSet.source);
+        auto tileId = *std::get<2>(loadTask);
+        auto& tileSet = *std::get<1>(loadTask);
+        auto tileIt = tileSet.tiles.find(tileId);
+        auto& entry = tileIt->second;
 
-        bool cached = tileSet.source->getTileData(task);
+        auto task = tileSet.source->createTask(tileId);
 
-        if (cached || m_loadPending < int(MAX_DOWNLOADS)) {
+        if (task->hasData()) {
+            // NB: Set implicit 'loading' state
+            entry.task = task;
+            m_dataCallback.func(std::move(task));
 
-            if (tile->sourceGeneration() < tileSet.source->generation()) {
-                // use new tile for updating while the stale tile
-                // is kept in tileSet for rendering.
-                tile->reloading = true;
-                task->tile = std::make_shared<Tile>(id, m_view->getMapProjection(),
-                                                    tileSet.source.get());
-            }
-
-            setTileState(*task->tile, TileState::loading);
-
-            if (cached) {
-                m_dataCallback.func(std::move(task));
-
-            } else if (!tileSet.source->loadTileData(std::move(task),
-                                                     m_dataCallback)) {
-                LOGE("Loading failed for tile [%d, %d, %d]", id.z, id.x, id.y);
-                m_loadPending--;
+        } else if (m_loadPending < int(MAX_DOWNLOADS)) {
+            entry.task = task;
+            if (tileSet.source->loadTileData(std::move(task), m_dataCallback)) {
+                m_loadPending++;
+            } else {
+                // TODO: How to handle this case?
```

Comment:
We can introduce a `tries` member to `DownloadTask`. And increment the tries. If it reaches a max, then we get rid of the task?